pkg/image/registryclient/client.go

```diff
@@ -352,19 +417,66 @@ func (c *Context) repositoryTransport(t http.RoundTripper, registry *url.URL, re
 	return c.cachedTransport(t, c.scopes(repoName))
 }
 
+func (c *Context) repoForRef(ctx context.Context, locator imagereference.DockerImageReference, altRef imagereference.DockerImageReference, insecure bool) (distribution.Repository, imagereference.DockerImageReference, error) {
```

Comment:
Why does this method need both `locator` and `altRef`?

Reply:
they are not always the same. The locator is the passed imagereference and altRef is the possible mirror alternative found from ICSP object. If you have an ICSP spec like this:
```
spec:
  repositoryDigestMirrors:
  - mirrors:
    - registry.ci.openshift.org/ocp/4.7
    source: does/not/exist
```
you can run something like
`oc image extract does/not/exist:tests --path=/:/tmp/ext  --icsp-file ~icsp.yaml` 
The locator: does/not/exist:tests
The altRef: registry.ci.openshift.org/ocp/4.7

Reply:
I'd expect altRef to be `registry.ci.openshift.org/ocp/4.7:tests`

Reply:
altRef is an ref.AsRepository() and the altRef:locator ID is tried

Reply:
I understand, but altRef could (and I think should) include all necessary information.

Reply:
but you can't know the digest, the alternates only tell you that the registry is accessible, for release images that reference multiple digests that can't be stored. (w/ oc adm release extract --command openshift-install, you _first_ need to know you can access the release image by its digest, then you can check for the digest of the installer - so the installer ID has to be attached to the accessible AsRepository() ref ) -anyways, clayton has suggested an extensive rewrite, so i'm working on that now